---
title: 'A paired-end ChiP-Seq pipeline to process paired-end reads'
tags:
  - Python
  - Snakemake
  - ChIP-seq
  - Deeptools

authors:
  - name: Jihed Chouaref
    orcid: 0000-0003-3865-896X
    affiliation: 1 # (Multiple affiliations must be quoted)
  - name: Mattijs Bliek
    orcid: 0000-0002-0488-4873
    affiliation: 1
    - name: Marc Galland
      orcid: 0000-0003-2161-8689 
      affiliation: 1
affiliations:
 - name: Swammerdam Institute for Life Sciences, University of Amsterdam
   index: 1
 - name: Institution 2
   index: 2
date: 14 december 2018
bibliography: paper.bib
---

# Summary

Chromatin immunoprecipitation followed by high-throughput sequencing (ChIP-seq) is a powerful tool for investigation the genome-wide distribution of DNA binding protein and their modification. The computational analysis of the great amount of data generated by this method is still a bottleneck for most of the researchers. Procedures like quality control, mapping raw data to the reference genome, peak calling, annotation and functional enrichment analysis requires various tools. These various tools require different software dependencies, variation in the software version might lead to unreproducible results.
Here we provide a user-friendly ChIP-seq analysis pipeline for paired-end data based on the Snakemake workflow manager. Only few modifications on the configuration files and to the tabular file containing sample information are required for the adaptability of the ChIP-seq pipeline to any organism, from which the genome have been sequenced and annotated.
The scalability and reproducibility of the data analysis is insured by Snakemake through creation and deployment of environments including the software dependencies using Conda or Singularity.
Raw Illumina paired-end data are processed by the pipeline and are subsequently trimmed, mapped and processed automatically according to the parameters set in the configuration file.
The outputs delivered by the pipeline are:

(1)	Quality controls files to check for the quality of the reads. Reads are processed by programs such as `fastp` and `deeptools` in order to produce graph that are easily readable and inform quickly about the quality of the experiment.

(2)	Portable visualization files for the observation of the read coverage on the genome using genome viewer.

(3)	Peaks informations files, these `bed` files gather the information about the peak calling produced by the MACS2 algorithm. This files can be potentially used for annotation and functional enrichment analysis.

(4)	The deeptools suite used by the pipeline produces beautiful visualization of the read coverage over genomic features provided by the user

This Snakemake ChIP-seq analysis pipeline provide an easy to use command line pipeline requiring minimum modifications with high modularity for domain knowledge input from the user. The first release of this pipeline has been archived to Zenodo with the folloing linked DOI : [![DOI](https://zenodo.org/badge/141444770.svg)](https://zenodo.org/badge/latestdoi/141444770)


# Usage

## Conda environment

An environment containing Snakemake is required and can installed using the [Conda package manager](https://conda.io/docs/using/envs.html).
1. Create a virtual environment named "chipseq" from the `global_env.yaml` file with the following command: `conda env create --name chipseq --file ~/envs/global_env.yaml`
2. Then, activate this virtual environment with `source activate chipseq`

The Snakefile will then take care of installing and loading the packages and softwares required by each step of the pipeline.

# Test run on small fastq test files

To test this pipeline, you will need Snakemake installed on your local machine. If you have activated the 'chipseq' environment (see above), then Snakemake 5.2.0 is already installed and in use.    

The repository contains subset of real ChIP-seq Illumina paired-end data, which can be used to ensure that the pipeline
operate properly.

If the `singularity` software is available on your machine and you want to use 10 CPUs (`--cores 10`), then run `snakemake --use-conda --use-singularity --cores 10`. Otherwise, run `snakemake --use-conda --cores 10`

# Citations

Citations to entries in paper.bib should be in
[rMarkdown](http://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html)
format.
To cite this pipeline, use:  
Chouaref et al., (2019). A paired-end ChiP-Seq pipeline to process paired-end reads.
Journal of Open Source Software, 3(31), 1102, https://doi.org/10.21105/joss.01102

# Directed Acyclic Graph of the Snakemake ChIP-seq PE pipeline

This graph has been produced with the command: `snakemake --rulegraph |dot -Tpng > dag.png`
![](dag.png)

# Acknowledgements

We acknowledge contributions from

# References

Köster, Johannes and Rahmann, Sven. “Snakemake - A scalable bioinformatics workflow engine”. Bioinformatics 2012.

Ramírez, Fidel, Devon P. Ryan, Björn Grüning, Vivek Bhardwaj, Fabian Kilpert, Andreas S. Richter, Steffen Heyne, Friederike Dündar, and Thomas Manke. “deepTools2: a next generation web server for deep-sequencing data analysis.” Nucleic Acids Research (2016): gkw257.
